crates/bevy_ecs/src/world/deferred_world.rs

```diff
@@ -303,15 +322,58 @@ impl<'w> DeferredWorld<'w> {
     #[inline]
     pub(crate) unsafe fn trigger_on_remove(
         &mut self,
+        archetype: &Archetype,
         entity: Entity,
         targets: impl Iterator<Item = ComponentId>,
     ) {
-        for component_id in targets {
-            // SAFETY: Caller ensures that these components exist
-            let hooks = unsafe { self.world.components().get_info_unchecked(component_id) }.hooks();
-            if let Some(hook) = hooks.on_remove {
-                hook(DeferredWorld { world: self.world }, entity, component_id);
+        if archetype.has_remove_hook() {
+            for component_id in targets {
+                let hooks =
+                // SAFETY: Caller ensures that these components exist
+                    unsafe { self.world.components().get_info_unchecked(component_id) }.hooks();
+                if let Some(hook) = hooks.on_remove {
+                    hook(DeferredWorld { world: self.world }, entity, component_id);
+                }
             }
         }
     }
+
+    /// Triggers all event observers for [`ComponentId`] in target.
+    ///
+    /// # Safety
+    /// Caller must ensure observers listening for `event` can accept ZST pointers
+    #[inline]
+    pub(crate) unsafe fn trigger_observers(
+        &mut self,
+        event: ComponentId,
+        source: Option<Entity>,
+        components: impl Iterator<Item = ComponentId>,
+    ) {
+        Observers::invoke(self.reborrow(), event, source, components, &mut ());
+    }
+
+    /// Triggers all event observers for [`ComponentId`] in target.
+    ///
+    /// # Safety
+    /// Caller must ensure `E` is accessible as the type represented by `event`
+    #[inline]
+    pub(crate) unsafe fn trigger_observers_with_data<E>(
+        &mut self,
+        event: ComponentId,
+        source: Option<Entity>,
+        components: impl Iterator<Item = ComponentId>,
+        data: &mut E,
+    ) {
+        Observers::invoke(self.reborrow(), event, source, components, data);
+    }
+
+    /// Constructs an [`EventBuilder`] for an ECS event.
+    pub fn ecs_event<E: Component>(&mut self, data: E) -> EventBuilder<E> {
+        EventBuilder::new(data, self.commands())
+    }
```

Comment:
It's there for completeness with the APIs on World and Commands, I don't have an example specifically using the deferred world implementation since it's the same as the other two.